### ISSUE TITLE
JupyterLab Running Panel: display kernels only once per instance - and show list of connected sessions

### ISSUE THREAD

**Issue by SylvainCorlay** [jupyterlab/jupyterlab#13779]
The "kernels" section of the "Running" left panel currently shows inconsistent data:

In the following screenshot, we see the state after opening a notebook and then connecting a console to it, which seems correct (only one kernel, two tabs):

<img width="252" alt="Screenshot 2023-01-16 at 09 46 07" src="https://user-images.githubusercontent.com/2397974/212635711-ff269217-cb25-4954-923d-027180fc2843.png">

In the following screenshot, we see the state after opening a console and then connecting a notebook to it, which seems incorrect (same kernel listed twice - killing one kills the other one):

<img width="248" alt="Screenshot 2023-01-16 at 09 48 49" src="https://user-images.githubusercontent.com/2397974/212635929-5478136b-ef31-454e-921a-1e75831cc351.png">

The proposed fix is to list all kernels by name in the kernels pane, and list the sessions:

<img width="248" alt="fixed" src="https://user-images.githubusercontent.com/2397974/212640605-a86c875d-a36c-4a32-ae1a-2e34100ae312.png">

As a second iteration, we could make that list of connected tabs expendable below the kernel name, but getting to that state would already be a proper bug fix.

**Comment by krassowski** [comment]
Could we iterate on the UX introduced in #13792; I find it to be a downgrade in the UX for the most common action the user takes (of closing the kernel they would like to close):

![Screenshot from 2023-01-24 10-35-45](https://user-images.githubusercontent.com/5832902/214270009-b9f843cb-efe7-45f3-9441-a28c88e5e80b.png)

1. In worst case scenario it requires user to hover over every kernel, say 20 Python 3 kernels for 20 open noteboks, to find the kernel they want to close
2. It uses Jupyter icon rather than Python icon (as proposed in the this issue). In fact I think it should use a stack of Notebook/Console icons (if the kernel is used by Notebook, only use Notebook, if it is used by Notebook and Console show both icons overlaid with small offset)
3. Minor: curly braces do not have an obvious meaning to me. Maybe we could overlay the number over the icon in a corner?
4. Minor: I don't think we need to show the number if it is equal to 1.

Thoughts @SylvainCorlay?

**Comment by afshin** [comment]
@krassowski 
I have been working on a follow-on PR to improve this UI based on #13792, because I agree that it isn't a good enough user experience.

- The logo of the kernel just happens to be very annoying to display with the current implementation of `LabIcon`, otherwise it would have gone in during the first pass.
- I think I agree with you about curly braces. I only chose them because we already had a pair of parentheses there, but two pairs is probably better than introducing a new concept.
- I've added a context menu to try to simplify things as a second iteration but not as the final UI. I will open in a new PR and will link it in this thread now that the CI gods have allowed the first iteration to go in. It looks like this:

https://user-images.githubusercontent.com/159529/214274436-16129ecc-4081-4d05-bd62-9c343cb0a04f.mov

But even this, I would say, is incomplete. I think it'd be best to refactor the running sidebar's kernel pane to allow a list of top-level kernel that are expandable (with a little `+`) to show the session connections beneath them. I think the `contextmenu` approach in the video above is an improvement, but I agree that seeing everything in one glance is better.

**Comment by jtpio** [comment]
> In worst case scenario it requires user to hover over every kernel, say 20 Python 3 kernels for 20 open noteboks, to find the kernel they want to close

Maybe it will be fixed by https://github.com/jupyterlab/jupyterlab/pull/13792, but it seems like with https://github.com/jupyterlab/jupyterlab/pull/13792 now merged it's more difficult to find what document a kernel refers to.

For example on the following screenshot (running on the latest `master`) I closed the notebook widget so it does not show up in the "Open Tabs" anymore, but there is a dangling "Python kernel" and it's hard to know where it came from:

https://user-images.githubusercontent.com/591645/214309039-244c7c4b-d9fa-4853-b3ec-683e732bc538.mp4

I personally liked the fact that before, the name of the associated notebook (or path) was displayed under the "Kernels" section. It made easier to quickly know from which notebook the kernel was started.

Can this be added back in https://github.com/jupyterlab/jupyterlab/pull/13851? Without having to right click or use the context menu, so it works better on mobile and in Notebook 7. Maybe as a (nested) list of paths under the kernel name?

**Comment by afshin** [comment]
> Maybe as a (nested) list of paths under the kernel name?

This will be part of the follow-on PR. The "dangling" kernel is actually a real live kernel session that still exists but had no visibility before. I think having a nest structure with 0 or more children for each kernel will more accurately reflect the state of your application as well as allow you to connect to kernels that come from other sources.

**Comment by jtpio** [comment]
> The "dangling" kernel is actually a real live kernel session that still exists but had no visibility before

What I meant is the following, the fact that users could know to which path a kernel was associated after closing the notebook widget in lab. In this example it days "Untitled.ipynb" which is more useful than "Python kernel":

https://user-images.githubusercontent.com/591645/214317835-acb53b9a-4a6b-48ba-94a2-ee602d412697.mp4

**Comment by afshin** [comment]
The trouble before was that if you had a console and a notebook both connected to the same kernel, they showed up as two items in the kernels list. Then if you clicked the `X` for the console, you would kill the kernel of your notebook and lose your state.

I believe that we can have both the correct behavior and also the full UI information if we switch to the nested UI.

**Comment by jtpio** [comment]
> I believe that we can have both the correct behavior and also the full UI information if we switch to the nested UI.

Yes that would be great! The proposed screenshot in the top comment (https://github.com/jupyterlab/jupyterlab/issues/13779#issue-1534532419) seems to be suggesting a similar approach by listing the console and notebooks in parenthesis. But a nested or collapsible list would probably be easier to read and interact with.

**Comment by krassowski** [comment]
Nested list sounds good to me. Maybe something like:

![Screenshot from 2023-01-24 18-04-04](https://user-images.githubusercontent.com/5832902/214373909-4d97b799-f5cb-483f-bff0-0e9603a196ff.png)

If rendering the kernel icon is prohibitively difficult (hopefully it is not) we could use a circle as sketched above - Jupyter users already have an association of the circular icon with kernel (regardless of the kernel logo) as it is used in busy/progress indicator). We could use the empty space to show the number of documents/widgets connected to the kernel. More than 9 are unlikely bur we could use `∞` or `9+` in that case.

> The "dangling" kernel is actually a real live kernel session that still exists but had no visibility before.

It was visible before too (and is on 3.x and 2.x). This is the where users refer to when they run low on memory (killing old kernels from closed notebooks that were previously closed). Maybe we even should have a button for "Kill all kernels which do not have any notebook/console".

On this use-case we could have something like this:

![Screenshot from 2023-01-24 19-53-59](https://user-images.githubusercontent.com/5832902/214395002-c1fbba37-d280-49d3-a28d-d233cd922843.png)

**Comment by fcollonval** [comment]
Following #13851, the panel looks now like that:

https://raw.githubusercontent.com/jupyterlab/jupyterlab/2e8677624ed7301d47de703570a14ce3bb8036cf/galata/test/documentation/general.test.ts-snapshots/running-layout-documentation-linux.png

Should we close this issue and open a follow-up issue for the _Close dangling_ feat. proposed by @krassowski ?
